### ISSUE TITLE
Failure to run training in AWS on Pyspacer Patch Model

### ISSUE THREAD

**Issue by SparkKitty** [data-mermaid/mermaid-classifier#17]
## Summary/Issue
Pyspacer training runs do not seem to be finishing, according to ML flow results.

## Impact/Priority
Blocking progress on running with large datasets. For example, I have been unable to complete a training run with all MERMAID annotations (221,983 annotations at last successful training run) with all the CoralNet source annotations that we used in the current MERMAID model (322 sources). 

## Testing Steps
I was able to complete training runs using an "ml.t3.large" Sagemaker instance with all MERMAID annotations plus the first 5 CoralNet sources (451,808 total annotations; run took <1 hour) and with all MERMAID annotations plus first 10 CoralNet sources (556,839 total annotations; run took 1.3 hours) but when I tried to run with MERMAID plus the first 20 CoralNet sources it did not finish even after 9 days (1,109,799 total annotations). As a test, I ran with the same dataset (MERMAID plus first 20 CoralNet sources) but using an XL instance ("ml.t3.xlarge") and it completed in 3.3 hours using that larger instance. I was able to use that XL instance to run with MERMAID annotations plus the first 40 CoralNet Sources (1,775,052 total annotations; run took 5.2 hours) but an even larger dataset (MERMAID plus the first 80 CoralNet sources) failed to complete even after 5 days. Based on this testing, I am guessing that using an even larger instance or changing to a different type of instance (e.g. ml.**m5**.xlarge instead of ml.**t3**.xlarge), but it would be better to determine the underlying issue so we can test with even more data without needing to use huge instances. 

## Request
Could someone please investigate the underlying issue for why large datasets could be failing to complete?

## Recommended Next Steps
I would recommend testing with a larger instance and a different type of instance (e.g. M5) to see if those are actually solutions for larger datasets but I would also recommend looking into the underlying code to see if it is likely a memory issue or something else (see notes from Stephen as well).

<img width="1424" height="664" alt="Image" src="https://github.com/user-attachments/assets/cb3065f3-2df5-4e5a-bf72-b2ed50231598" />

**Comment by saanobhaai** [comment]
Two ideas for diagnosis from Stephen:

- The reference set memory calculation might not be doing what we want; one idea is to try a run using a conservative value (overriding the default memory-based calculation) to see if it finishes; if it does this would be a place to intervene.
- One pyspacer optimization Stephen introduced last year was reusing a boto session for repeated image access requests; for a long-running task, it's possible the boto session is timing out.

**Comment by StephenChan** [comment]
> * One pyspacer optimization Stephen introduced last year was reusing a boto session for repeated image access requests; for a long-running task, it's possible the boto session is timing out.

I just happened to be running a test for this on CoralNet, for similar setups - reusing a boto session that was initiated through IMDS rather than a manually specified AWS access key. I reused after over 12 hours, and the session still seemed to be good. 12 hours is the longest max session length that I could find in AWS docs, although FWIW, it was difficult to confirm which maximum applied to this particular setup.

**Comment by gnieuwenhuis** [comment]
I'll keep updating as I go, but as a first datapoint:

With `SPACER_REF_SET_MAX_SIZE=30000` (more conservative), it took about 3 hours to run MERMAID plus the first 20 CoralNet sources (1108799 annotations) on `ml.t3.large` (this failed after 9 days with default ref setting).

This suggest that this is a memory issue, but I want to test the larger one as well, and then come up with a plan for fixing.

**Comment by gnieuwenhuis** [comment]
After some hunting and looking at Iain's previous test results, here's a summary of what I've determined and some thoughts on next steps:

## Observations:
instance | memory | annotations | ref_set_max_size | Execution time
-- | -- | -- | -- | --
ml.t3.large | 8gb | 1.1 million | 30,000 | 3 hours
ml.m5.large (non-burstable) | 8gb | 1.1 million | 50,000 (default) | 2.5 hours
ml.t3.xlarge | 16gb | 1.1 million | 100,000 | 2.8 hours

- Lowering the `ref_set_max_size` I was able to successfully run the 1.1 million annotation experiment within 8GB of RAM. This same experiment previously failed to complete after 9 days using the default value of `ref_set_max_size`. 
- Changing the instance type from t3 (burstable) to m5 allowed that same experiment to run successfully with the default value of `ref_set_max_size`
- Increasing the `ref_set_max_size` and increasing the instance size to 16GB RAM resulted in a successful run as well, even with bigger batch and ref dataset size, but it took longer to run.


## Interpretation

1. Lowering the memory pressure allowed experiments to run that otherwise did not, and so did changing to non-burstable instance type. This suggests that memory thrashing via swap is the most likely cause (because it did not OOM, it just continued without faiing), and CPU throttling caused by a burstable instance type is (at minimum) aggravating the problem.
2. Increasing the `ref_set_max_size` within a larger instance size did not complete faster as I would have expected with a larger batch size. I believe this can be explained by the larger ref set size, causing pre-epoch validation and calibration to take longer, which is not entirely offset by the speed up caused by larger training batch sizes in this case.

## Problem Definition

#### Current state:

The ref set is used for:
- calculating per-epoch accuracy
- calibrating probabilities after training is finished

The ref set also:
- is held in memory
- it's size is capped by the training batch size even if the defined split would include more.

Training batch size (TRAINING_BATCH_LABEL_COUNT) comes directly from SPACER_REF_SET_MAX_SIZE, and this value also caps `automatic_ref_max_size()` which defines the actual number of rows in the reference dataset.

#### Problems with the current state:

1. For labels with very few annotations, capping the ref_set like this can lead to poor calibration of those rare labels (I have no confirmation this is happening, it's just a potential risk).
2. Increasing the batch size means two separate datasets in memory are affected. There is no way to increase batch size while holding the ref set the same, or vice versa. Since these two datasets are both held in memory, this limits the options one can take when memory problems do occur.
3. [minor] the naming scheme is confusing - SPACER_REF_SET_MAX_SIZE is mapped directly to TRAINING_BATCH_LABEL_COUNT introducing some misdirection.


## Possible Solutions

1. Decouple ref dataset size from training batch size, and validate & calibrate in batches from disk to avoid having the reference dataset in memory at all (preferred solution).
2. Docs only change: give instructions for how to set SPACER_REF_SET_MAX_SIZE whenever too much memory is being taken up during training.
3. Improve the automatic_ref_max_size() calculation to avoid memory thrashing via swap.

Improvement for any solution: memory logging for future diagnosis.

**Comment by gnieuwenhuis** [comment]
@StephenChan I'm most interested in your thoughts on this - especially what you think about solutions. I would like to go with solution 1 just for the general flexibility that provides, but I could see a simpler fix being possible in `automatic_ref_max_size()`. I'm not sure your level of involvement, but I have time to chat about it if you do.

**Comment by StephenChan** [comment]
@gnieuwenhuis Great analysis, thanks. I hadn't considered that burstable instances could be a factor at all, and that's interesting about the relationship between ref set size and training speed.

On decoupling ref set max size and batch size: I'm onboard with making these separate settings in pyspacer. As you point out, the memory pressure is a function of the two numbers, and one might want to make the ref set bigger than the batch size to optimize this situation.

> ... validate & calibrate in batches from disk to avoid having the reference dataset in memory at all

I'm curious what this would look like though. Since this would mean changing the code that's very close to the scikit-learn function calls (and since my core knowledge of training algorithms is limited), I'm not currently confident enough to implement this in a way that keeps the spirit of the original algorithm.

> Improvement for any solution: memory logging for future diagnosis.

I'm also onboard with this and had been thinking about it already. There should be at least one logging message available per batch so that it's easier to tell if the training run is still going, and including the memory usage in the message would make sense.

Also agree with adding more docs guidance on this issue overall. And open to any suggestions on improving the automatic_ref_max_size() calculation.

**Comment by gnieuwenhuis** [comment]
I'll see if I can put together a skeleton example showing what might be necessary to validate and calibrate in batches from disk, I'm not 100% sure its possible with the scikit-learn setup, but I think it is.

**Comment by gnieuwenhuis** [comment]
@StephenChan I did some testing work which you can see in [this PR](https://github.com/data-mermaid/mermaid-classifier/pull/18).

I think this provides a path forward that removes the cap on the reference dataset, allows batching of all memory bound calculations (training, validation, and calibration), but it does introduce a little complexity, so I am curious your thoughts.

If you ignore the fact that I rewrote the train_loop from pyspacer in that PR (it was only done so I could experiment), take a look at the `_calibrate_in_batches` function. This is a re-implementation of the CalibratedClassifierCV from sklearn which allows batching. For the most part it is still using sklearn's core logic for calibration through the private method, but this new function allows making predictions in batches, which was the part of that algorithm that is memory bound.

I am in favor of implementing this to solve the overall memory problem, but I am unsure how we should split the logic between pyspacer and the classifier. In many ways having the train_loop defined in the classifier makes it much easier for me to experiment and do some of the work I am planning to do in the coming weeks, but I don't have your history on this project, and am unsure if that is an acceptable path forward.

**Comment by StephenChan** [comment]
@gnieuwenhuis I just got done looking through this, very interesting.

The contents of _calibrate_in_batches() make sense to me as I compare them with CalibratedClassifierCV.fit(), and I appreciate the helpful code comments here (as well as elsewhere). The tests you wrote to compare the training results between this experiment and current pyspacer are exactly what I was hoping for, too.

You're right that this adds complexity though, as well as some brittleness since it has to match up with the implementation details of CalibratedClassifierCV and pyspacer's ClassifierUnpickler, both of which may change later.

One pyspacer design point that's been on my mind sometimes: the existence of the ClassifierTrainer and MiniBatchTrainer classes in train_classifier.py seem to signal some intent to make the training logic customizable, but the potential is untapped. There isn't any meaningful inheritance going on between those two classes, and `MiniBatchTrainer.__call__()` calls a module-level function `train()` to do most of the work.

So from that thought, I wonder if the trainer class and function designs in pyspacer could be adjusted so that the _calibrate_in_batches() logic, as well as the per-epoch logging logic, could be defined in a subclass of ClassifierTrainer. And then the trainer_name parameter of TrainClassifierMsg could be reworked so that it can accept trainer classes defined outside of pyspacer, maybe similar to the `extractor` parameter of ExtractFeaturesMsg.

If that could be done, I think that'd be quite a reasonable way to draw the line between pyspacer and mermaid-classifier here, while hopefully enabling m-c to implement what it needs with a minimum of code duplication.

(Also, I did not know about mlflow's base_metrics_monitor; this seems like a super good use of it.)

**Comment by gnieuwenhuis** [comment]
That sounds like an ideal way forward to me - I think your plan will give me the freedom I need to experiment with the mermaid-classifier, while keeping a clean interface between pyspacer and the classifier.

If you agree, I can throw together an implementation of this in both pyspacer and m-c and we can iterate on it from there?

**Comment by gnieuwenhuis** [comment]
@StephenChan I just realized that I will need permission to push a branch to pyspacer. I do have some changes I sketched up that allow inheritance of the ClassifierTrainer in pyspacer, if you can give me permission I can push the branch and make a PR for you to look at.

**Comment by StephenChan** [comment]
@gnieuwenhuis Are you familiar with making a PR using a fork? I think that would be preferable here, since as far as I know, giving permission to push a branch to a GitHub repo is tied to broad 'Write' permissions to everything in that repo (managing all branches, issues, PRs, etc. except the protected `main` branch).

Note that in the fork case, there's still an extra step needed if there's a need to let upstream maintainers (i.e. me) add commits to the PR as well: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork

**Comment by gnieuwenhuis** [comment]
Ahh, interestingly I have never actually had to do that, but it works for me - PR is here: https://github.com/coralnet/pyspacer/pull/137

It's still in draft because I want to make sure I can use these changes in the mermaid-classifier before declaring it ready for review. I'll keep you posted.

**Comment by gnieuwenhuis** [comment]
@StephenChan I've marked the PR ready for review: https://github.com/coralnet/pyspacer/pull/137

I tested those changes in the mermaid-classifier [here](https://github.com/data-mermaid/mermaid-classifier/pull/18) successfully.

Let me know what you think - happy to iterate on it as needed.
